pkg/storage/unified/sql/data/resource_version_inc.sql

```diff
@@ -1,6 +1,6 @@
 UPDATE {{ .Ident "resource_version" }}
 SET
-    {{ .Ident "resource_version" }} = {{ .Arg .ResourceVersion}}
+    {{ .Ident "resource_version" }} = {{ .CurrentEpoch }}
```

Comment:
If we bring back the old logic of getting the current resourceversion and incrementing it, we could change this to something like `GREATEST({{ .CurrentEpoch }}, {{ .Arg .ResourceVersion }})` to "handle" the case where we have multiple entries in the same microsecond by running ahead to the next microsecond, on the theory that time will catch up before it returns anyway.

Reply:
Good idea. I used MAX instead of GREATEST because the later isn't compatible with sqlite.

Reply:
If memory serves, we needed to lock the row first to prevent the scenario where multiple updates happen concurrently and then the subsequent select receives the incorrect value. @diegommm can probably explain it more clearly than I am.

Reply:
Yes. That was the case when incrementing. But now that we are using a timestamp, it's pretty unlikely to endup with the same RV.

Reply:
I had to go back to using GREATEST in mysql+postgres because MAX doesn't support fetching a row record.

Reply:
@DanCech is right, we need to lock the row with a SELECT ... FOR UPDATE first. If we don't do that, writers will just race to UPDATE and will get dirty reads regardless of the db engine and isolation level (even serializable isolation level won't protect here), and eventual inconsistency is guaranteed. SQL is very awkward, it gives querying operations more power than to data modifying operations.

Reply:
Cool, so we can just revert to the select + update + select method and get that protection at the cost of 1 more db round trip

Reply:
That's correct. I'm not too worried about adding one more roundtrip here, though. They are roughly worth a few millis, our bottleneck is in the global row locking itself.

Reply:
Sure. Let's add the round trip then. This is really optimising the scenario of what happens on concurrent events when the clock goes backward ... which is very rare.

![image](https://github.com/user-attachments/assets/205c5973-127b-42a3-b9f5-aecb7db74e4c)